### PR TITLE
ClusterInfo string_view interface and abseil hashtables

### PR DESCRIPTION
**Pull request by MBkkt**
### Scope & Purpose

ClusterInfo string_view interface and abseil hashtables

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [x] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*
  - [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR: https://github.com/arangodb/enterprise/pull/890
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/Cluster/ClusterMethods.cpp [hunk 1]
```diff
@@ -767,8 +761,11 @@ CloneShardDistribution(ClusterInfo& ci, std::shared_ptr<LogicalCollection> col,
   for (uint64_t i = 0; i < numberOfShards; ++i) {
     // determine responsible server(s)
     std::string shardId = "s" + StringUtils::itoa(id + i);
-    result->try_emplace(std::move(shardId),
-                        otherShardsMap->at(otherShards.at(i)));
+    auto it = otherShardsMap->find(otherShards.at(i));
+    if (it == otherShardsMap->end()) {
+      throw std::runtime_error{""};
```
Comment (gnusi) [inline arangod/Cluster/ClusterMethods.cpp:766]: please specify an error message

File: arangod/Cluster/ClusterMethods.cpp [hunk 2]
```diff
@@ -3536,7 +3532,12 @@ arangodb::Result hotRestoreCoordinator(ClusterFeature& feature,
     // Check timestamps of all dbservers:
     size_t good = 0;  // Count restarted servers
     for (auto const& dbs : dbServers) {
-      if (postServersKnown.at(dbs) != preServersKnown.at(dbs)) {
+      auto it1 = postServersKnown.find(dbs);
+      auto it2 = preServersKnown.find(dbs);
+      if (it1 == postServersKnown.end() || it2 == preServersKnown.end()) {
+        throw std::runtime_error{""};
```
Comment (gnusi) [inline arangod/Cluster/ClusterMethods.cpp:3538]: please specify an error message

File: arangod/Cluster/Maintenance.cpp [hunk 3]
```diff
@@ -910,11 +912,15 @@ arangodb::Result arangodb::maintenance::diffPlanLocal(
     auto const collectLogInformation = [&] {
       auto const& localLogsInDatabase = localLogsByDatabase.at(dbname);
       auto planLogsInDatabase = ReplicatedLogSpecMap{};
+      auto it = plan.find(dbname);
+      if (it == plan.end()) {
+        throw std::runtime_error{""};
```
Comment (gnusi) [inline arangod/Cluster/Maintenance.cpp:917]: please specify an error message

File: arangod/Cluster/ClusterInfo.h [hunk 4]
```diff
@@ -269,9 +270,9 @@ class CollectionInfoCurrent {
   /// @brief local helper to return a map to boolean
   //////////////////////////////////////////////////////////////////////////////
 
-  template<typename T>
-  std::unordered_map<ShardID, bool> getFlag(T const& name) const {
-    std::unordered_map<ShardID, bool> m;
+  [[nodiscard]] containers::FlatHashMap<ShardID, bool> getFlag(
+      std::string const& name) const {
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.h:274]: Can we use `std::string_view` for `name` as well here?

File: arangod/Cluster/ClusterInfo.cpp [hunk 5]
```diff
@@ -215,23 +214,22 @@ void addToShardStatistics(
...
   return arangodb::AgencyOperation{
       "Plan/Version", arangodb::AgencySimpleOperationType::INCREMENT_OP};
 }
 
-static inline std::string collectionPath(std::string const& dbName,
-                                         std::string const& collection) {
-  return "Plan/Collections/" + dbName + "/" + collection;
+inline std::string collectionPath(std::string_view dbName,
+                                  std::string_view collection) {
+  return "Plan/Collections/" + std::string{dbName} + "/" + collection.data();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:224]:
```suggestion
  return "Plan/Collections/" + std::string{dbName} + "/" + std::string{collection};
```

File: arangod/Cluster/ClusterInfo.cpp [hunk 6]
```diff
@@ -5796,44 +5797,44 @@ void ClusterInfo::setFailedServers(
...
+bool ClusterInfo::serverExists(std::string_view serverId) const noexcept {
   READ_LOCKER(readLocker, _serversProt.lock);
-  return _servers.find(serverId) != _servers.end();
+  return _servers.contains(serverId);
 }
 
-bool ClusterInfo::serverAliasExists(std::string const& alias) const noexcept {
+bool ClusterInfo::serverAliasExists(std::string_view alias) const noexcept {
   READ_LOCKER(readLocker, _serversProt.lock);
   return _serverAliases.find(alias) != _serverAliases.end();
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:5825]:
```suggestion
  return _serverAliases.contains(alias);
```

File: arangod/Cluster/ClusterInfo.cpp [hunk 7]
```diff
@@ -6293,14 +6293,14 @@ ClusterInfo::ServersKnown::ServersKnown(
   }
 }
 
-std::unordered_map<ServerID, ClusterInfo::ServersKnown::KnownServer> const&
+containers::FlatHashMap<ServerID, ClusterInfo::ServersKnown::KnownServer> const&
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:6296]: It looks like this method is not called from anywhere, so we can as well remove it.

File: arangod/Cluster/ClusterInfo.cpp [hunk 8]
```diff
@@ -6426,7 +6424,7 @@ ClusterInfo::SyncerThread::SyncerThread(Server& server,
 
 ClusterInfo::SyncerThread::~SyncerThread() { shutdown(); }
 
-bool ClusterInfo::SyncerThread::notify(velocypack::Slice const& slice) {
+bool ClusterInfo::SyncerThread::notify(velocypack::Slice const&) {
```
Comment (jsteemann) [inline arangod/Cluster/ClusterInfo.cpp:6427]: Looks like we can change the signature here, in the header and at the call site.
```suggestion
bool ClusterInfo::SyncerThread::notify() {
```

File: arangod/Cluster/ClusterInfo.h [hunk 9]
```diff
@@ -248,15 +249,15 @@ class CollectionInfoCurrent {
...
+  [[nodiscard]] uint64_t getCurrentVersion() const { return _currentVersion; }
 
   //////////////////////////////////////////////////////////////////////////////
   /// @brief local helper to return boolean flags
   //////////////////////////////////////////////////////////////////////////////
 
  private:
-  template<typename T>
-  bool getFlag(T const& name, ShardID const& shardID) const {
+  [[nodiscard]] bool getFlag(std::string const& name,
```
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Cluster/ClusterInfo.h:259]: We use this function with StaticStrings, and here we use getBooleanValue that works with the const ref template. so I think const string ok here

File: arangod/Cluster/ClusterInfo.h (hunk 4 above)
Comment (MBkkt) [reply arangod/Cluster/ClusterInfo.h:274]: Same here

File: arangod/Cluster/ClusterInfo.cpp (hunk 5 above)
Comment (MBkkt) [reply arangod/Cluster/ClusterInfo.cpp:224]: No real reason. But ok

File: arangod/Cluster/ClusterInfo.cpp (hunk 8 above)
Comment (MBkkt) [reply arangod/Cluster/ClusterInfo.cpp:6427]: Ok

File: arangod/Cluster/ClusterInfo.cpp (hunk 7 above)
Comment (MBkkt) [reply arangod/Cluster/ClusterInfo.cpp:6296]: I will try

File: arangod/Cluster/ClusterInfo.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (MBkkt) [reply arangod/Cluster/ClusterInfo.h:259]: But I will try. Maybe in the future, we replace StaticString to constexpr std::string_view

**Review by jsteemann — APPROVED** [review]
LGTM!

**Pull request merged**